llvm/lib/Transforms/IPO/SampleProfile.cpp

```diff
@@ -444,15 +444,36 @@ class SampleProfileMatcher {
   // the profile.
   StringMap<LocToLocMap> FuncMappings;
 
-  // Profile mismatching statstics.
+  // Match state for an anchor/callsite.
+  enum class MatchState {
+    Matched = 0,
+    Mismatched = 1,
+    // Stay Matched after profile matching.
+    StayMatched = 2,
+    // Recovered from Mismatched after profile matching.
+    Recovered = 3,
+    Unknown = 32,
+  };
+
+  // For each function, store every callsite and its matching state into this
+  // map, of which each entry is a pair of callsite location and MatchState.
+  // This is used for profile staleness computation and report.
+  StringMap<std::unordered_map<LineLocation, MatchState, LineLocationHash>>
```

Comment:
note that StringMap keeps a copy of every string key.

Reply:
I see.  If we have the memory pressure, another option is to consolidate the two StringMap

```
StringMap<LocToLocMap> FuncMappings;
StringMap<std::unordered_map<LineLocation, MatchState, LineLocationHash>> FuncCallsiteMatchStates;
```

Into one map, like

```
StringMap<std::pair<LocToLocMap, std::unordered_map<LineLocation, MatchState, LineLocationHash>>>    FuncXXX... 
```
the Key of `LocToLocMap` is also a LineLocation, I was thinking to further merge them
```
StringMap<std::unordered_map<LineLocation, std::pair<LineLocation, MatchState>, LineLocationHash>>>    FuncXXX... 
```

but unfortunately the key of `LocToLocMap` is `IRlocation`  but the latter's key is the `ProfileLocation`, so merging them might be too hacky and confusing.

Reply:
> note that StringMap keeps a copy of every string key.

I think one optimization is we can release this data after the matching is done, we don't use those data for the later sample inlining or loading stuffs, similar to the `StringMap<LocToLocMap> FuncMappings;`.